Source/DafnyCore/Options/WarnMissingConstructorParenthesisOption.cs

```diff
@@ -0,0 +1,14 @@
+namespace Microsoft.Dafny;
+
+class WarnMissingConstructorParenthesisOption : BooleanOption {
+  public static readonly WarnMissingConstructorParenthesisOption Instance = new();
+  public override string LongName => "warn-missing-constructor-parentheses";
+  public override string Description => @"
+    Emits a warning when a constructor name in a case pattern is not
+    followed by parentheses.".TrimStart();
```

Comment:
I would prefer to not port the individual warning options to the new CLI.  Instead we should have a single option `--warn` taking a list of warning identifiers.  WDYT?

Reply:
So instead of `--warn-X --warn-Y --warn-Z`, users would write `--warn=+X+Y+Z`, or are you suggesting something else? It doesn't seem particularly easier to use and it adds a new concept to the CLI UI. Having separate options for separate warnings also makes it easier to define a help description for each of them.

Reply:
That's right, some syntax like this.  It's fine if the syntax and perhaps the implementation ends up looking like `--warn-xyz`, as long as there are easy ways to say:

- Turn on all warnings, except `X`
- Treat all warnings as errors, except `X` (treat it as a warning)
- Treat `X` as an error

etc (there may be some combinations I'm forgetting right now).  All of these are pretty standard features, I think (e.g. https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings).  Additionally, having identifiers for warnings and errors will make it easy to have docs (and explainers, down the line) for each of them (e.g. https://eslint.org/docs/latest/rules/getter-return or https://pylint.pycqa.org/en/latest/user_guide/messages/error/bad-format-character.html)

Reply:
> Treat all warnings as errors, except X
> Treat X as an error

What's the use-case for either of those options? To me a warning is a problem that must to be resolved before before merging to main, but not before development testing. We have the option `--warnings-as-errors` which should always be turned on during CI. Why would you want some warnings as errors and others not? Would you use that during development or CI ? During CI seems wrong because then you enable checking in warnings and your project may grow in amount of warnings leading you to ignore them. During development I can imagine but it seems too far fetched. It would be you differentiate between three categories of warnings:

1. Never useful, turn off
1. Too dangerous to be checked in, error in CI. Not too dangerous for exploratory testing, warning during development.
1. So dangerous you don't even want to do exploratory testing with this, since it's likely you have a bug.

Is there really a user that will differentiate between 2 and 3 ?

> Treat all warnings as errors, except X

I don't think anyone would use that option. Warnings that are off by default are the ones for which we think the extra safety is likely not worth the effort, and there will be an arbitrarily large bucket of such warnings. Is there any chance you would want approximately everything in the bucket of not recommended things? Also, if you use this option, upgrading to a newer Dafny version may prevent your project from compiling because a new warning has been added.

> Additionally, having identifiers for warnings and errors will make it easy to have docs

Why? It'll be harder to have CLI documentation for it.

---

How about we introduce the `--warn-X` syntax now, and we leave the option open to have `--error-X` in the future, and `--all-warnings` together with `--warn-X=false` ?

Reply:
I think in the context of this PR, just supporting `--warn-X` for now is fine.

In a wider scope, though, I think having these warnings enabled by default and being able to disable individual warnings (probably even specific instances of warnings on specific bits of code) is really important for them to actually be useful.

Should we be considering enabling warnings by default in Dafny 4?

Reply:
> What's the use-case for either of those options?

Same for other languages: precise control about warnings.
In our case, we have warnings that are informational with high false-positive rates, and others that are true issues.  Different teams will want to handle them differently.
For example, the non-determinism flags should be warnings, not a separate option.  That warning could be on by default.  Some teams would turn it off; other would turn it (selectively) into an error.

> Why? It'll be harder to have CLI documentation for it.

For the reason I listed above: we can have one webpage per warning identifier.
I don't understand how having unique identifiers for warnings and errors will make CLI documentation harder (we can still have individual CLI flags, though I'm not sold on that being useful).

> Also, if you use this option, upgrading to a newer Dafny version may prevent your project from compiling because a new warning has been added.

So does `--warnings-as-errors`, which you say should always be turned on during CI, right?
In fact, having a way to treat all warnings as errors except for a few specific exception makes this problem easier to handle: if you have a codebase with `warnings-as-errors` and we add a warning in a new Dafny release, your code will break (assuming you trigger the warning). If you have a flag that allows to turn off warnings as errors for that specific warning, you can push that and still benefit from the update. With the current situation, you have to turn off warnings-as-error for *all* warnings.

Reply:
> if you have a codebase with warnings-as-errors and we add a warning [that is on by default] in a new Dafny release, your code will break (assuming you trigger the warning)

If we add a check that's on by default, then we need to add an option to turn that check off to be backwards compatible.

> Same for other languages: precise control about warnings.
In our case, we have warnings that are informational with high false-positive rates, and others that are true issues. Different teams will want to handle them differently.

My argument was that you only need to be able to turn the check fully on or fully off, where fully on would mean it's a warning during development and an error during CI.

I think it's generally considered bad practice to have something that doesn't prevent you from running the program, to be an error during development. It's important to be able to code freely, run and test, and only then go into polishing your program so it passes CI.

> For example, the non-determinism flags should be warnings, not a separate option. 

I agree that the difference between language feature flags and checks doesn't seem so crisp!